packages/xo-server/src/xo-mixins/proxies.js

```diff
@@ -0,0 +1,64 @@
+import { noSuchObject } from 'xo-common/api-errors'
+
+import Collection from '../collection/redis'
+import patch from '../patch'
+
+export default class Proxy {
+  constructor(app) {
+    const db = (this._db = new Collection({
+      connection: app._redis,
+      prefix: 'xo:proxy',
+    }))
+
+    app.on('clean', () => db.rebuildIndexes())
+    app.on('start', () =>
+      app.addConfigManager(
+        'proxies',
+        () => db.get(),
+        proxies => db.update(proxies)
+      )
+    )
+  }
+
+  async registerProxyAppliance({ address, authenticationToken, name }) {
+    const proxies = await this.getAllProxies()
+    const proxy = proxies.find(proxy => proxy.address === address)
+    if (proxy !== undefined) {
+      throw new Error(
+        `A proxy (${proxy.id}) with the address (${
+          proxy.address
+        }) is already registered.`
+      )
+    }
+
+    const { properties } = await this._db.add({
+      address,
+      authenticationToken,
+      name,
+    })
+    return properties
+  }
```

Comment:
This method needs to be synchronized to avoid race conditions.

Race condition:

```js
await Promise.all([
  xo.registerProxyAppliance({ address: '192.168.1.1' }),
  xo.registerProxyAppliance({ address: '192.168.1.1' }),
])
```